src/kafka.cpp

```diff
@@ -0,0 +1,252 @@
+    throw;
+  }
+
+  // XXX ugly duplication
+  rdk_producer_for_statuses.reset(RdKafka::Producer::create(conf, errstr));
+  if (!rdk_producer_for_statuses) {
+      print::debug("KafkaProducer: failed to create Producer (for statuses)");
+    throw;
+  }
+
+  host_name = boost::asio::ip::host_name();
+
+  print::debug("KafkaProducer load_config done");
+}
+
+void KafkaProducer::report_failure(const std::string& site, const std::string& ip) {
+  json message;
+  message["id"] = host_name;
+  message["name"] = "ip_failed_challenge";
+  message["value"] = ip;
+  std::string serialized_message = message.dump();
+  size_t serialized_message_size = message.dump().size();
+
+  RdKafka::ErrorCode err = rdk_producer_for_failed_challenges->produce(
+      /* Topic name */
+      failed_challenge_topic,
+      /* Any Partition: the builtin partitioner will be
+              * used to assign the message to a topic based
+              * on the message key, or random partition if
+              * the key is not set. */
+      (int)RdKafka::Topic::PARTITION_UA,
+      /* Make a copy of the value */
+      (int)RdKafka::Producer::RK_MSG_COPY /* Copy payload */,
+      /* Value */
+      (void*)serialized_message.c_str(), serialized_message_size,
+      /* Key */
+      nullptr, 0,
+      /* Timestamp (defaults to current time) */
+      0,
+      /* Message headers, if any */
+      NULL);
+  if (err == RdKafka::ERR_NO_ERROR) {
+      print::debug("reported challenge failure for site: ", site, " and ip: ", ip, " to topic: ", failed_challenge_topic);
+  } else {
+      print::debug(": Failed to produce()");
+  }
+}
+
+void KafkaProducer::report_status() {
+  json message;
+  message["id"] = host_name;
+
+  std::string serialized_message = message.dump();
+  size_t serialized_message_size = message.dump().size();
+
+  RdKafka::ErrorCode err = rdk_producer_for_statuses->produce(
+      /* Topic name */
+      status_topic,
+      /* Any Partition: the builtin partitioner will be
+              * used to assign the message to a topic based
+              * on the message key, or random partition if
+              * the key is not set. */
+      (int)RdKafka::Topic::PARTITION_UA,
+      /* Make a copy of the value */
+      (int)RdKafka::Producer::RK_MSG_COPY /* Copy payload */,
+      /* Value */
+      (void*)serialized_message.c_str(), serialized_message_size,
+      /* Key */
+      nullptr, 0,
+      /* Timestamp (defaults to current time) */
+      0,
+      /* Message headers, if any */
+      NULL);
+  if (err == RdKafka::ERR_NO_ERROR) {
+      print::debug("reported status");
+  } else {
+      print::debug(": Failed to report status()");
+  }
+}
+
+void
+KafkaConsumer::reload_config(YAML::Node& config, Banjax* banjax) {
+    TSMutexLock(stored_config_lock);
+    auto on_scope_exit = defer([&] { TSMutexUnlock(stored_config_lock); });
+    stored_config = config;
+    banjax = banjax;
+    config_valid = false;
```

Comment:
I don't understand here as well. If we support the hot switch of config, then we should re-create the thread again. I mean it might be new topic names for example.